test/filters/test_crop_hull.cpp

```diff
@@ -0,0 +1,495 @@
+        insideMask(inputCloud->size(), false),
+        insideIndices(insideIndices),
+        insideCloud(new pcl::PointCloud<pcl::PointXYZ>),
+        outsideCloud(new pcl::PointCloud<pcl::PointXYZ>)
+    {
+      pcl::copyPointCloud(*inputCloud, insideIndices, *insideCloud);
+      for (pcl::index_t idx : insideIndices) {
+        insideMask[idx] = true;
+      }
+      for (size_t i = 0; i < inputCloud->size(); ++i) {
+        if (!insideMask[i]) {
+          outsideIndices.push_back(i);
+        }
+      }
+      pcl::copyPointCloud(*inputCloud, outsideIndices, *outsideCloud);
+    }
+
+    pcl::PointCloud<pcl::PointXYZ>::ConstPtr inputCloud;
+    std::vector<bool> insideMask;
+    pcl::Indices insideIndices, outsideIndices;
+    pcl::PointCloud<pcl::PointXYZ>::Ptr insideCloud, outsideCloud;
+  };
+
+
+  std::vector<TestData>
+  createTestDataSuite(
+      std::function<pcl::PointXYZ()> insidePointGenerator,
+      std::function<pcl::PointXYZ()> outsidePointGenerator)
+  {
+    std::vector<TestData> testDataSuite;
+    size_t const chunkSize = 1000;
+    pcl::PointCloud<pcl::PointXYZ>::Ptr insideCloud(new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::PointCloud<pcl::PointXYZ>::Ptr outsideCloud(new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::PointCloud<pcl::PointXYZ>::Ptr mixedCloud(new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::Indices insideIndicesForInsideCloud;
+    pcl::Indices insideIndicesForOutsideCloud; // empty indices, cause outsideCloud don't contains any inside point
+    pcl::Indices insideIndicesForMixedCloud;
+    for (size_t i = 0; i < chunkSize; ++i)
+    {
+      insideIndicesForInsideCloud.push_back(i);
+      insideCloud->push_back(insidePointGenerator());
+      outsideCloud->push_back(outsidePointGenerator());
+      if (randomBool()) {
+        insideIndicesForMixedCloud.push_back(i);
+        mixedCloud->push_back(insidePointGenerator());
+      }
+      else {
+        mixedCloud->push_back(outsidePointGenerator());
+      }
+    }
+    testDataSuite.emplace_back(std::move(insideIndicesForInsideCloud), insideCloud);
+    testDataSuite.emplace_back(std::move(insideIndicesForOutsideCloud), outsideCloud);
+    testDataSuite.emplace_back(std::move(insideIndicesForMixedCloud), mixedCloud);
+    return testDataSuite;
+  }
+
+
+  class PCLCropHullTestFixtureBase : public ::testing::Test
+  {
+    public:
+      PCLCropHullTestFixtureBase()
+        : gen(12345u),
+          rd(0.0f, 1.0f)
+      {
+        baseOffsetList.emplace_back(0, 0, 0);
+        baseOffsetList.emplace_back(5, 1, 10);
+        baseOffsetList.emplace_back(1, 5, 10);
+        baseOffsetList.emplace_back(1, 10, 5);
+        baseOffsetList.emplace_back(10, 1, 5);
+        baseOffsetList.emplace_back(10, 5, 1);
+      }
+    protected:
+      virtual pcl::PointCloud<pcl::PointXYZ>::ConstPtr getBaseInputCloud() const = 0;
+      virtual int getDim() const = 0;
+
+      void test(std::function<void(pcl::CropHull<pcl::PointXYZ>, TestData const &)> checker) const
```

Comment:
Why didn't you use `SetUp` here? Make `testDataSuite` a class member and then every test will have access to it. You can then have the checker run in the test itself instead of inverting control

Reply:
The problem in `baseOffsetList`.
For every `baseOffset` need individual `(inside|outside)PointGenerator`. I don't have any idea, how this function could be refactor.

Reply:
`SetUp` is run after ctor but before the actual test. So L160-L164 can be separate if each `TestData` is generated and stored (in a vector in the class) using the `SetUp` before the test runs

Reply:
Basically, split the function based on generation part and testing part

Reply:
splitted `test` function to `SetUp/TearDown` and execution loops in every test case